File: interop/go/physicalir/geometry.go
package physicalir

import (
	"fmt"
	"math"
	"math/big"
)

func integer(value int64) *big.Int {
	return big.NewInt(value)
}

func difference(left, right int64) *big.Int {
	return new(big.Int).Sub(integer(left), integer(right))
}

func cross(a, b, c Point) *big.Int {
	abX := difference(b.X, a.X)
	abY := difference(b.Y, a.Y)
	acX := difference(c.X, a.X)
	acY := difference(c.Y, a.Y)
	left := new(big.Int).Mul(abX, acY)
	right := new(big.Int).Mul(abY, acX)
	return new(big.Int).Sub(left, right)
}

func between(a, b, point Point) bool {
	return min(a.X, b.X) <= point.X && point.X <= max(a.X, b.X) &&
		min(a.Y, b.Y) <= point.Y && point.Y <= max(a.Y, b.Y)
}

func segmentsIntersect(a, b, c, d Point) bool {
	abC := cross(a, b, c).Sign()
	abD := cross(a, b, d).Sign()
	cdA := cross(c, d, a).Sign()
	cdB := cross(c, d, b).Sign()
	if abC*abD < 0 && cdA*cdB < 0 {
		return true
	}
	return abC == 0 && between(a, b, c) ||
		abD == 0 && between(a, b, d) ||
		cdA == 0 && between(c, d, a) ||
		cdB == 0 && between(c, d, b)
}

func signedDoubleArea(vertices []Point) *big.Int {
	total := new(big.Int)
	if len(vertices) == 0 {
		return total
	}
	for index, a := range vertices {
		b := vertices[(index+1)%len(vertices)]
		left := new(big.Int).Mul(integer(a.X), integer(b.Y))
		right := new(big.Int).Mul(integer(b.X), integer(a.Y))
		total.Add(total, new(big.Int).Sub(left, right))
	}
	return total
}

func polygonContains(vertices []Point, point Point, boundary bool) bool {
	winding := 0
	for index, a := range vertices {
		b := vertices[(index+1)%len(vertices)]
		turn := cross(a, b, point).Sign()
		if turn == 0 && between(a, b, point) {
			return boundary
		}
		if a.Y <= point.Y && point.Y < b.Y && turn > 0 {
			winding++
		} else if b.Y <= point.Y && point.Y < a.Y && turn < 0 {
			winding--
		}
	}
	return winding != 0
}

func validateRectangle(rectangle Rectangle, location string) error {
	if rectangle.Width <= 0 || rectangle.Height <= 0 {
		return validationError(location, "rectangle dimensions must be positive")
	}
	right := new(big.Int).Add(integer(rectangle.Origin.X), integer(rectangle.Width))
	top := new(big.Int).Add(integer(rectangle.Origin.Y), integer(rectangle.Height))
	if !right.IsInt64() || !top.IsInt64() {
		return validationError(location, "rectangle edge exceeds signed 64-bit range")
	}
	return nil
}

func validatePolygon(polygon Polygon, location string) error {
	vertices := polygon.Vertices
	if len(vertices) < 3 || len(vertices) > MaxPolygonVertices {
		return validationError(location, "polygon requires 3..%d vertices", MaxPolygonVertices)
	}
	seen := make(map[Point]struct{}, len(vertices))
	for index, point := range vertices {
		if _, exists := seen[point]; exists {
			return validationError(location, "polygon vertex %d is duplicated", index)
		}
		seen[point] = struct{}{}
	}
	for index, a := range vertices {
		b := vertices[(index+1)%len(vertices)]
		c := vertices[(index+2)%len(vertices)]
		if cross(a, b, c).Sign() == 0 && !between(a, c, b) {
			return validationError(location, "adjacent polygon edges backtrack at vertex %d", index+1)
		}
		for other := index + 2; other < len(vertices); other++ {
			if index == 0 && other == len(vertices)-1 {
				continue
			}
			if segmentsIntersect(a, b, vertices[other], vertices[(other+1)%len(vertices)]) {
				return validationError(location, "polygon edges %d and %d intersect non-adjacently", index, other)
			}
		}
	}
	if signedDoubleArea(vertices).Sign() == 0 {
		return validationError(location, "polygon area must be non-zero")
	}
	return nil
}

func validatePath(path Path, location string) error {
	if len(path.Points) < 2 || len(path.Points) > MaxPathPoints {
		return validationError(location, "path requires 2..%d points", MaxPathPoints)
	}
	if path.Width <= 0 {
		return validationError(location, "path width must be positive")
	}
	minimumX, maximumX := path.Points[0].X, path.Points[0].X
	minimumY, maximumY := path.Points[0].Y, path.Points[0].Y
	for index, point := range path.Points {
		if index > 0 && point == path.Points[index-1] {
			return validationError(location, "consecutive path points %d and %d are equal", index-1, index)
		}
		minimumX, maximumX = min(minimumX, point.X), max(maximumX, point.X)
		minimumY, maximumY = min(minimumY, point.Y), max(maximumY, point.Y)
	}
	width := integer(path.Width)
	twiceMinimum := integer(math.MinInt64)
	twiceMinimum.Mul(twiceMinimum, big.NewInt(2))
	twiceMaximum := integer(math.MaxInt64)
	twiceMaximum.Mul(twiceMaximum, big.NewInt(2))
	left := new(big.Int).Sub(new(big.Int).Mul(integer(minimumX), big.NewInt(2)), width)
	bottom := new(big.Int).Sub(new(big.Int).Mul(integer(minimumY), big.NewInt(2)), width)
	right := new(big.Int).Add(new(big.Int).Mul(integer(maximumX), big.NewInt(2)), width)
	top := new(big.Int).Add(new(big.Int).Mul(integer(maximumY), big.NewInt(2)), width)
	if left.Cmp(twiceMinimum) < 0 || bottom.Cmp(twiceMinimum) < 0 ||
		right.Cmp(twiceMaximum) > 0 || top.Cmp(twiceMaximum) > 0 {
		return validationError(location, "path envelope exceeds signed 64-bit range")
	}
	return nil
}

func validateGeometry(geometry Geometry, location string) error {
	switch geometry.Kind {
	case RectangleKind:
		if geometry.Rectangle == nil || geometry.Polygon != nil || geometry.Path != nil {
			return validationError(location, "rectangle geometry union is inconsistent")
		}
		return validateRectangle(*geometry.Rectangle, location)
	case PolygonKind:
		if geometry.Rectangle != nil || geometry.Polygon == nil || geometry.Path != nil {
			return validationError(location, "polygon geometry union is inconsistent")
		}
		return validatePolygon(*geometry.Polygon, location)
	case PathKind:
		if geometry.Rectangle != nil || geometry.Polygon != nil || geometry.Path == nil {
			return validationError(location, "path geometry union is inconsistent")
		}
		return validatePath(*geometry.Path, location)
	default:
		return validationError(location, "unknown geometry kind %q", geometry.Kind)
	}
}

func geometryCost(geometry Geometry) (int, int, error) {
	switch geometry.Kind {
	case RectangleKind:
		if geometry.Rectangle == nil {
			return 0, 0, fmt.Errorf("rectangle payload is absent")
		}
		return 4, 16, nil
	case PolygonKind:
		if geometry.Polygon == nil {
			return 0, 0, fmt.Errorf("polygon payload is absent")
		}
		count := len(geometry.Polygon.Vertices)
		return count, count * count, nil
	case PathKind:
		if geometry.Path == nil {
			return 0, 0, fmt.Errorf("path payload is absent")
		}
		return len(geometry.Path.Points), 0, nil
	default:
		return 0, 0, fmt.Errorf("unknown geometry kind %q", geometry.Kind)
	}
}

File: src/spicetrellis/geometry.py
"""Exact database-unit geometry for physical views.

Coordinates are signed 64-bit integers. Bounds may be half-integral for odd-width
paths. No floating-point tolerance changes connectivity, area or orientation.
"""

from __future__ import annotations

from dataclasses import dataclass
from fractions import Fraction

INT64_MIN = -(2**63)
INT64_MAX = 2**63 - 1
MAX_POLYGON_VERTICES = 1_024
MAX_PATH_POINTS = 100_000


class GeometryError(ValueError):
    """Invalid, unsupported or out-of-range database-unit geometry."""


def coordinate(value: int, name: str = "coordinate") -> int:
    """Require an actual integer in the wire contract, never bool or float."""
    if type(value) is not int or not INT64_MIN <= value <= INT64_MAX:
        raise GeometryError(f"{name} must be a signed 64-bit integer")
    return value


def _positive(value: int, name: str) -> None:
    coordinate(value, name)
    if value <= 0:
        raise GeometryError(f"{name} must be positive")


@dataclass(frozen=True, slots=True)
class Point:
    x: int
    y: int

    def __post_init__(self) -> None:
        coordinate(self.x, "x")
        coordinate(self.y, "y")


@dataclass(frozen=True, slots=True)
class Bounds:
    """Closed signed64-range bounds on the half-unit grid, including path envelopes."""

    left: Fraction | int
    bottom: Fraction | int
    right: Fraction | int
    top: Fraction | int

    def __post_init__(self) -> None:
        for value in (self.left, self.bottom, self.right, self.top):
            if type(value) not in (int, Fraction):
                raise GeometryError("bounds require exact integers or fractions")
            if value < INT64_MIN or value > INT64_MAX:
                raise GeometryError("bounds exceed signed 64-bit coordinate range")
            if isinstance(value, Fraction) and value.denominator not in (1, 2):
                raise GeometryError("bounds must lie on the integer or half-unit grid")
        if self.left > self.right or self.bottom > self.top:
            raise GeometryError("bounds must be ordered")

    def union(self, other: Bounds) -> Bounds:
        if type(other) is not Bounds:
            raise GeometryError("union requires Bounds")
        return Bounds(
            min(self.left, other.left),
            min(self.bottom, other.bottom),
            max(self.right, other.right),
            max(self.top, other.top),
        )

    def contains(self, point: Point) -> bool:
        if type(point) is not Point:
            raise GeometryError("containment requires a Point")
        return self.left <= point.x <= self.right and self.bottom <= point.y <= self.top


def _point_tuple(points: tuple[Point, ...], minimum: int, maximum: int) -> None:
    if type(points) is not tuple or not minimum <= len(points) <= maximum:
        raise GeometryError(f"points must be an immutable tuple of {minimum}..{maximum} points")
    if any(type(point) is not Point for point in points):
        raise GeometryError("each vertex must be a Point")


def _bounds(points: tuple[Point, ...]) -> Bounds:
    return Bounds(
        min(point.x for point in points),
        min(point.y for point in points),
        max(point.x for point in points),
        max(point.y for point in points),
    )


def _cross(a: Point, b: Point, c: Point) -> int:
    return (b.x - a.x) * (c.y - a.y) - (b.y - a.y) * (c.x - a.x)


def _between(a: Point, b: Point, point: Point) -> bool:
    return min(a.x, b.x) <= point.x <= max(a.x, b.x) and min(a.y, b.y) <= point.y <= max(a.y, b.y)


def _intersects(a: Point, b: Point, c: Point, d: Point) -> bool:
    ab_c, ab_d, cd_a, cd_b = _cross(a, b, c), _cross(a, b, d), _cross(c, d, a), _cross(c, d, b)
    if (ab_c > 0 > ab_d or ab_d > 0 > ab_c) and (cd_a > 0 > cd_b or cd_b > 0 > cd_a):
        return True
    return (
        (ab_c == 0 and _between(a, b, c))
        or (ab_d == 0 and _between(a, b, d))
        or (cd_a == 0 and _between(c, d, a))
        or (cd_b == 0 and _between(c, d, b))
    )


def _orient(point: Point, clockwise: int, reflect_x: bool) -> tuple[int, int]:
    x, y = point.x, -point.y if reflect_x else point.y
    if clockwise == 90:
        return y, -x
    if clockwise == 180:
        return -x, -y
    if clockwise == 270:
        return -y, x
    return x, y


@dataclass(frozen=True, slots=True)
class Polygon:
    """A simple polygon without holes; the closing vertex is implicit.

    Validation uses at most 1,024 vertices and quadratic exact segment tests.
    Clockwise and counter-clockwise vertex order are both retained losslessly.
    """

    vertices: tuple[Point, ...]

    def __post_init__(self) -> None:
        _point_tuple(self.vertices, 3, MAX_POLYGON_VERTICES)
        if len(set(self.vertices)) != len(self.vertices):
            raise GeometryError("polygon vertices must be distinct; closure is implicit")
        count = len(self.vertices)
        for index, a in enumerate(self.vertices):
            b, c = self.vertices[(index + 1) % count], self.vertices[(index + 2) % count]
            if _cross(a, b, c) == 0 and not _between(a, c, b):
                raise GeometryError("adjacent polygon edges may not backtrack")
            for other in range(index + 2, count):
                if index == 0 and other == count - 1:
                    continue  # Adjacent closing edge shares the first vertex.
                if _intersects(a, b, self.vertices[other], self.vertices[(other + 1) % count]):
                    raise GeometryError("polygon edges may not intersect or touch non-adjacently")
        if self.signed_double_area == 0:
            raise GeometryError("polygon must have non-zero area")

    @property
    def signed_double_area(self) -> int:
        return sum(
            a.x * b.y - b.x * a.y
            for a, b in zip(self.vertices, (*self.vertices[1:], self.vertices[0]), strict=True)
        )

    @property
    def area(self) -> Fraction:
        return Fraction(abs(self.signed_double_area), 2)

    @property
    def bounds(self) -> Bounds:
        return _bounds(self.vertices)

    def contains(self, point: Point, *, boundary: bool = True) -> bool:
        """Exact winding-number containment, with an explicit boundary policy."""
        if type(point) is not Point or type(boundary) is not bool:
            raise GeometryError("containment needs a Point and boolean boundary policy")
        winding = 0
        for a, b in zip(self.vertices, (*self.vertices[1:], self.vertices[0]), strict=True):
            cross = _cross(a, b, point)
            if cross == 0 and _between(a, b, point):
                return boundary
            if a.y <= point.y < b.y and cross > 0:
                winding += 1
            elif b.y <= point.y < a.y and cross < 0:
                winding -= 1
        return winding != 0


@dataclass(frozen=True, slots=True)
class Rectangle:
    origin: Point
    width: int
    height: int

    def __post_init__(self) -> None:
        if type(self.origin) is not Point:
            raise GeometryError("rectangle origin must be a Point")
        _positive(self.width, "width")
        _positive(self.height, "height")
        coordinate(self.origin.x + self.width, "rectangle right edge")
        coordinate(self.origin.y + self.height, "rectangle top edge")

    @property
    def area(self) -> int:
        return self.width * self.height

    @property
    def bounds(self) -> Bounds:
        return Bounds(
            self.origin.x, self.origin.y, self.origin.x + self.width, self.origin.y + self.height
        )

    def polygon(self) -> Polygon:
        x, y = self.origin.x, self.origin.y
        return Polygon(
            (
                Point(x, y),
                Point(x + self.width, y),
                Point(x + self.width, y + self.height),
                Point(x, y + self.height),
            )
        )


@dataclass(frozen=True, slots=True)
class Path:
    """Constant-width centerline; its bounds conservatively include square caps.

    Cap/join polygonization is not implicit. These bounds are not a fabricated
    area, DRC clearance, or Boolean union of segments.
    """

    points: tuple[Point, ...]
    width: int

    def __post_init__(self) -> None:
        _point_tuple(self.points, 2, MAX_PATH_POINTS)
        _positive(self.width, "path width")
        if any(a == b for a, b in zip(self.points, self.points[1:], strict=False)):
            raise GeometryError("consecutive path points must differ")
        # Bounds validates half-grid endpoints, including the stroke envelope.
        _ = self.bounds

    @property
    def bounds(self) -> Bounds:
        center = _bounds(self.points)
        half = Fraction(self.width, 2)
        return Bounds(
            center.left - half, center.bottom - half, center.right + half, center.top + half
        )


@dataclass(frozen=True, slots=True)
class Transform:
    """Reflect in the local x-axis, rotate clockwise, then translate.

    The eight Manhattan orientations are exact integer isometries. Arbitrary
    rotations are rejected, rather than rounded onto a different physical grid.
    Composition follows function application: ``a.compose(b)(p) == a(b(p))``.
    """

    origin: Point = Point(0, 0)
    clockwise: int = 0
    reflect_x: bool = False

    def __post_init__(self) -> None:
        if type(self.origin) is not Point or type(self.reflect_x) is not bool:
            raise GeometryError("transform requires a Point origin and boolean reflection")
        if type(self.clockwise) is not int or self.clockwise not in (0, 90, 180, 270):
            raise GeometryError("only exact 0/90/180/270 degree clockwise rotations are supported")

    def apply(self, point: Point) -> Point:
        if type(point) is not Point:
            raise GeometryError("transform input must be a Point")
        # Do not range-check the intermediate point: translation may bring a
        # reflected INT64_MIN coordinate back into the representable domain.
        x, y = _orient(point, self.clockwise, self.reflect_x)
        return Point(x + self.origin.x, y + self.origin.y)

    def compose(self, inner: Transform) -> Transform:
        if type(inner) is not Transform:
            raise GeometryError("composition requires a Transform")
        rotation = (
            self.clockwise + (-inner.clockwise if self.reflect_x else inner.clockwise)
        ) % 360
        return Transform(self.apply(inner.origin), rotation, self.reflect_x != inner.reflect_x)

    def inverse(self) -> Transform:
        rotation = self.clockwise if self.reflect_x else (-self.clockwise) % 360
        # Negation and the inverse linear map must be evaluated before checking
        # the final range; e.g. a reflection can make -INT64_MIN representable.
        x, y = _orient(self.origin, rotation, self.reflect_x)
        return Transform(Point(-x, -y), rotation, self.reflect_x)

    def shape(self, shape: Rectangle | Polygon | Path) -> Polygon | Path:
        if type(shape) is Rectangle:
            x, y = shape.origin.x, shape.origin.y
            return Polygon(
                tuple(
                    self.apply(point)
                    for point in (
                        Point(x, y),
                        Point(x + shape.width, y),
                        Point(x + shape.width, y + shape.height),
                        Point(x, y + shape.height),
                    )
                )
            )
        if type(shape) is Polygon:
            return Polygon(tuple(self.apply(point) for point in shape.vertices))
        if type(shape) is Path:
            return Path(tuple(self.apply(point) for point in shape.points), shape.width)
        raise GeometryError("unsupported shape type")
